### DISCUSSION TITLE
Is it possible to know which patients had records produced by which modules?

### DISCUSSION THREAD

**Discussion by fabkury** [synthetichealth/synthea#1246] [Q&A] [answered]
My understanding is that Synthea runs each patient through all modules, and that most (maybe all) modules are probabilistic, which means each patient has some chance to "pass unscathed" by a module. That is, not ever have the condition of the module.

Is it possible to obtain the complete list of all patients who had records produced by each module?

For example, if module X gives each patient a 2% chance of being diagnosed with a given cancer, is it possible to obtain the list of those 2% somewhere? Without checking for the presence of the diagnoses codes/lab tests/synthetic patient data itself. I'm looking for some sort of external, authoritative, "ground truth" about the synthetic patients.

Thanks!

**Comment by fabkury** [comment]
I guess my question could also be phrased as:

Is it possible to establish (with certainty) from which state, of which module, came each row of data in the CSVs that Sythea produces?

Thanks!

**Comment by fabkury** [comment]
I realized the JSON export page (https://github.com/synthetichealth/synthea/wiki/JSON-Export) describes an option to export each patient's path through each module. Just what I was looking for.

**Comment by lhs-open** [comment]
The Json export record probably does have the module information directly. How do you tell which record or which patient is a result of which module?  That's your original question, right?

**Comment by fabkury** [comment, marked as answer]
You're right, lhs-open.

I was just able to export each patient's module history (using option `exporter.json.include_module_history`). I see that it gives you the list of states the patient went through, but **not** what records in the output were produced from that state (which is what I had asked for).

**So I guess the correct answer to my question is: _unfortunately no :/_**

This would be a great feature to have for purposes of learning electronic patient phenotyping. The module history is like the "answer key" to the exercise. The exercise is to identify patients with a given condition by querying the available patient data.

I _might_ be able to make do with just the list of states from each module, but here's my +1 for this feature to get implemented at some point: provide an option that lets you know, for each record in the output, exactly which state of which module produced it.

Thanks for all.

**Reply by annayudovin** [reply]
I've been wrestling with a similar problem, and since, up to this point I've been concentrating exclusively on CSV exports, it never even occurred to me to look into this JSON option (but thanks for pointing it out - I was about ready to try exploring other formats, and I may as well start with JSON).

I did, however, come up with a quick-and-dirty alternative solution for this. Even if you aren't trying to create a custom module (and this tactic is particularly useful for testing a custom module you are working on), the Synthea Module Builder (https://synthetichealth.github.io/module-builder) can load any of the "stock" modules, which gives you the ability to "tweak" them to your heart's content, download the modified copy, and use to generate the records. You probably already guessed where I'm going with this, right? While not all module states generate a record (this is especially true for CSVs), the ones that do often have attributes that can be modified with some unique text without affecting the logic. 

For instance, the "ConditionOnset" state's "Code" attribute refers (usually) to real SNOMED codes, which may also be of interest to other modules (there are conditional transitions that may check if a patient has a condition referred to by such a code). However, the "Display" attribute need not contain (or, at least, be strictly limited to) the "real" description of the SNOMED code. If you append a marker string of your choice there, and run your simulation, identifying patients affected by this "marked" condition will be a simple matter of a text search (even if other modules can cause the same condition, it will not have the same "customized" name!)

In fact, in some cases, you may even find it useful to add your own, custom, record-generating state(s) just prior to the one you are interested in where the transition to the target state is re-routed to the new state, but otherwise unchanged, and the new state features a "simple" direct transition to the target state - but this may be well beyond the scope of what you need.

This is, fundamentally, a hack - and a very inelegant one, so it rankles me to let it be a part of anything like a polished, presentable solution ( for that, I plan to look into custom exporter capabilities shortly), but it can make all the difference if you are trying to find out something about a specific state of a specific module.

**Reply by fabkury** [reply]
I'm not sure if I understand all you explained, but thanks a lot for your detailed thoughts! I think it would be great to have the detailed history, it would enable many use cases of the data.

**Reply by annayudovin** [reply]
If this is something you are still wrestling with, I think I can create an example to illustrate what I'm talking about - it's really not as complicated as it sounds (and if you'd rather do this via email, try ayudovin at gmail). No worries if you've already moved on to something else :)

**Reply by ofajardo** [reply]
hi @annayudovin I am interested on looking at an example, and I am sure it would be useful for others as well. Would you mind to share?

**Reply by fabkury** [reply]
@annayudovin now that I re-read I understood what you explained. Thanks for writing it out.

What you said requires rebuilding the dataset, but can work in some use cases for sure.

My use case is more nuanced because I am looking at the Synthea data through an ETL'd version of it. My Synthea data is in OMOP CDM format. So, to go with your strategy, I would need to figure out how to make the custom identifiers/hacks appear on the other side of the ETL (and not harm the ETL).

While I appreciate @annayudovin 's idea very much, I maintain my vote for this to be implemented in Synthea: full ability to trace exactly which state of which module generated each record in the Synthea output.

**Reply by annayudovin** [reply]
@fabkury: I certainly agree with you, the ability to trace exactly which state of which module generated each record in the Synthea output would be extremely useful (and more widely applicable than my hack)

In the meantime, however, if @ofajardo (or anyone else) is interested in how to implement said hack, I'm attaching the module I modified along with a step-by-step description of what I did to put it to use.
[cystic_fibrosis.json](https://github.com/user-attachments/files/16681957/cystic_fibrosis.json)

I opened the stock version of the Cystic Fibrosis module in Synthea Module Builder at https://synthetichealth.github.io/module-builder and modified the first ConditionOnset state by selecting the "Display" in the "Codes" section, and appending string "####" at the end. 
![CFmodule](https://github.com/user-attachments/assets/3b1f9ea4-d404-42e1-afea-f79f21c2c7cb)
![CFstateEdit](https://github.com/user-attachments/assets/eb8da6d0-f49e-4ad2-8297-ee9c166e0ff4)
I then downloaded the modified module and copied it into ~\synthea\src\main\resources\modules directory, overwriting the previous version of cystic_fibrosis.json.

In the .json file the change looks like this:
"codes": [
        {
          "system": "SNOMED-CT",
          "code": 190905008,
          "display": "Cystic Fibrosis ####"
        }

I replaced the synthea.properties file (in ~\synthea\bin\main\ directory) with the "stock" version (downloaded from github), because the one I've been using was heavily altered, rolled back the other customizations to my build and rebuilt the executables, then ran Synthea with the following string: 

.\run_synthea.bat -p 1000 -s 999 -cs 999 Massachusetts Boston
(meaning generate 1000 patients in Boston, using 999 as seed and 999 as clinician seed)

I then searched in ~\synthea\output directory for the added string "####", which showed up in one file - not very surprising, since cystic fibrosis is a rare disease.
